### PR TITLE
First draft of safehttp godoc comments.

### PR DESCRIPTION
**Pull request by kele**
Relevant: #9.

### CODE REVIEW

File: safehttp/doc.go [hunk 1]
```diff
@@ -12,5 +12,113 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Package safehttp TODO(@kele, @empijei)
+// Package safehttp provides a framework for building secure-by-default HTTP
+// servers. See https://github.com/google/go-safeweb#readme to learn about the
```
Comment (empijei) [inline safehttp/doc.go:16]:
```suggestion
// Package safehttp provides a framework for building secure-by-default web
// applications. See https://github.com/google/go-safeweb#readme to learn about the
```

File: safehttp/doc.go [hunk 2]
```diff
@@ -12,5 +12,113 @@
...
+// Safe Responses
+//
+// HTTP responses need to be crafted carefully in order to prevent common web
+// vulnerabilities like Cross-site Scripting (XSS). To help with this, we use
+// Safe Responses.
+//
+// Safe Responses are HTTP responses which has been determined to be safe when
+// received by a modern, popular web browser. For example, we consider HTML
+// responses generated using a template system enforcing contextual autoescaping
+// to be safe e.g. modern Angular, github.com/google/safehtml/template).
```
Comment (empijei) [inline safehttp/doc.go:28]:
```suggestion
// to be safe e.g. modern Angular or github.com/google/safehtml/template).
```

File: safehttp/doc.go [hunk 3]
```diff
@@ -12,5 +12,113 @@
...
+//
+// Safe Responses
+//
+// HTTP responses need to be crafted carefully in order to prevent common web
+// vulnerabilities like Cross-site Scripting (XSS). To help with this, we use
+// Safe Responses.
+//
+// Safe Responses are HTTP responses which has been determined to be safe when
+// received by a modern, popular web browser. For example, we consider HTML
+// responses generated using a template system enforcing contextual autoescaping
```
Comment (empijei) [inline safehttp/doc.go:27]: maybe link to https://rawgit.com/mikesamuel/sanitized-jquery-templates/trunk/safetemplate.html or https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/42934.pdf or https://blogtitle.github.io/go-safe-html/ or something like that to give a reference on what the term "contextual autoescaping" means.

File: safehttp/doc.go [hunk 4]
```diff
@@ -12,5 +12,113 @@
...
+// to be safe e.g. modern Angular, github.com/google/safehtml/template).
+//
+// The Go Safe Web ResponseWriter implementation accepts exclusively Safe
+// Responses, instead of byte slices.
```
Comment (empijei) [inline safehttp/doc.go:31]:
```suggestion
// Responses instead of byte slices.
```

File: safehttp/doc.go [hunk 5]
```diff
@@ -12,5 +12,113 @@
...
+// Dispatcher
+//
+// An implementation of a Dispatcher should be provided by security experts in
+// your project. The Dispatcher is called for every write method of the
+// ResponseWriter and is used to determine whether the passed response should be
+// considered safe. The Dispatcher is responsible for writing the response to
+// the underlying http.ResponseWriter in a safe way.
+//
+// Go Safe Web provides a DefaultDispatcher implementation which supports
+// github.com/google/safehtml responses.
```
Comment (empijei) [inline safehttp/doc.go:46]: welp, not just those...

File: safehttp/doc.go [hunk 6]
```diff
@@ -12,5 +12,113 @@
...
+// servers. See https://github.com/google/go-safeweb#readme to learn about the
+// goals and features.
+//
+// Safe Responses
+//
+// HTTP responses need to be crafted carefully in order to prevent common web
+// vulnerabilities like Cross-site Scripting (XSS). To help with this, we use
+// Safe Responses.
+//
+// Safe Responses are HTTP responses which has been determined to be safe when
```
Comment (bsiegert) [inline safehttp/doc.go:25]: nit: have been determined

File: safehttp/doc.go [hunk 7]
```diff
@@ -12,5 +12,113 @@
...
+// safe, we offer a way of configuring this in the framework. Whether a response
+// is considered safe or not is determined by the Dispatcher.
+//
+// Dispatcher
+//
+// An implementation of a Dispatcher should be provided by security experts in
+// your project. The Dispatcher is called for every write method of the
+// ResponseWriter and is used to determine whether the passed response should be
+// considered safe. The Dispatcher is responsible for writing the response to
+// the underlying http.ResponseWriter in a safe way.
```
Comment (empijei) [inline safehttp/doc.go:43]: I'd highlight that the Dispatcher might contain vulnerabilities and that the security of the overall project is relying on a sound Dispatcher implementation

File: safehttp/doc.go (hunk 2 above)
Comment (bsiegert) [inline safehttp/doc.go:28]: nit: looks like there is an open paren missing before e.g.

File: safehttp/doc.go [hunk 8]
```diff
@@ -12,5 +12,113 @@
...
+// Interceptors
+//
+// Not all security features can be implemented using the Dispatcher alone. For
+// instance, some requests should be rejected before they reach the handler in
+// order to prevent from Cross-site Request Forgery (CSRF, XSRF). To support
+// this, the framework uses Interceptors.
+//
+// An Interceptor implements methods that run before the request is passed to
+// the handler, and after the handler has committed a response. These are,
+// respectively, Before and After.
```
Comment (empijei) [inline safehttp/doc.go:57]: After? You mean Commit?

File: safehttp/doc.go [hunk 9]
```diff
@@ -12,5 +12,113 @@
...
+// HTTP responses need to be crafted carefully in order to prevent common web
+// vulnerabilities like Cross-site Scripting (XSS). To help with this, we use
+// Safe Responses.
+//
+// Safe Responses are HTTP responses which has been determined to be safe when
+// received by a modern, popular web browser. For example, we consider HTML
+// responses generated using a template system enforcing contextual autoescaping
+// to be safe e.g. modern Angular, github.com/google/safehtml/template).
+//
+// The Go Safe Web ResponseWriter implementation accepts exclusively Safe
```
Comment (bsiegert) [inline safehttp/doc.go:30]: s/accepts exclusively/only accepts/

File: safehttp/doc.go [hunk 10]
```diff
@@ -12,5 +12,113 @@
...
+// Life of a Request
+//
+// To tie it all together, we will explain how a single request goes through the
+// framework.
+//
+// When the request reaches the server, the following happens:
+//
+// 1. The ServeMux routes the request (i.e. picks the appropriate Handler that
+// will be eventually called). The HTTP method of the request is checked whether
+// it matches any registered Handler.
```
Comment (empijei) [inline safehttp/doc.go:68]:
```suggestion
// will be eventually called). The HTTP method and path of the request are checked against registered Handlers.
```

File: safehttp/doc.go [hunk 11]
```diff
@@ -12,5 +12,113 @@
...
+//
+// 2. [Before phase] The request is passed to all installed Interceptors, to
+// their Before methods, in the order of installation on the ServeMux. Each of
```
Comment (empijei) [inline safehttp/doc.go:71]:
```suggestion
// 2. [Before phase] The request is passed to all installed Interceptors via
// their Before methods, in the order of installation on the ServeMux. Each of
```

File: safehttp/doc.go [hunk 12]
```diff
@@ -12,5 +12,113 @@
...
+// To tie it all together, we will explain how a single request goes through the
+// framework.
+//
+// When the request reaches the server, the following happens:
+//
+// 1. The ServeMux routes the request (i.e. picks the appropriate Handler that
+// will be eventually called). The HTTP method of the request is checked whether
+// it matches any registered Handler.
+//
+// 2. [Before phase] The request is passed to all installed Interceptors, to
```
Comment (bsiegert) [inline safehttp/doc.go:70]: The requestis passed to the Before methods of all installed Interceptors in the order of their installation on the ServeMux.

File: safehttp/doc.go [hunk 13]
```diff
@@ -12,5 +12,113 @@
...
+// of subsequent Interceptors.
+//
+// 3. The request is passed to the Handler. The Handler calls a write method of
+// the ResponseWriter (e.g. Write, WriteError, Redirect...).
+//
+// 5. [Commit phase] Commit methods of the installed Interceptors are called, in
+// an inverted order (i.e. first Interceptor to be called in Before phase is
+// called last in the Commit phase). Commit methods can no longer influence the
+// flow of the request, but can still set headers, cookies, response HTTP status
+// code or even modify the response (if it's type allows it).
```
Comment (empijei) [inline safehttp/doc.go:84]:
```suggestion
// code or even modify the response (if its type allows it).
```

File: safehttp/doc.go (hunk 8 above)
Comment (bsiegert) [inline safehttp/doc.go:57]: Before, Commit and After?

File: safehttp/doc.go [hunk 14]
```diff
@@ -12,5 +12,113 @@
...
+//
+// 6. [Dispatcher] After all Commit methods have been called, the framework
+// passes the request to the Dispatcher. The Dispatcher determine the
```
Comment (empijei) [inline safehttp/doc.go:87]:
```suggestion
// passes the request to the Dispatcher. The Dispatcher determines the
```

File: safehttp/doc.go [hunk 15]
```diff
@@ -12,5 +12,113 @@
...
+// Content-Type of the response and whether the response should be considered a
+// Safe Response. It is responsible for writing the response to the underlying
```
Comment (empijei) [inline safehttp/doc.go:89]:
```suggestion
// Safe Response. The Dispatcher is responsible for writing the response to the underlying
```

File: safehttp/doc.go [hunk 16]
```diff
@@ -12,5 +12,113 @@
...
+// the Before methods can either let the execution continue by returning
+// safehttp.NotWritten() (and not using any ResponseWriter write methods), or by
+// actually writing a response. This would prevent further Before method calls
+// of subsequent Interceptors.
+//
+// 3. The request is passed to the Handler. The Handler calls a write method of
+// the ResponseWriter (e.g. Write, WriteError, Redirect...).
+//
+// 5. [Commit phase] Commit methods of the installed Interceptors are called, in
+// an inverted order (i.e. first Interceptor to be called in Before phase is
```
Comment (bsiegert) [inline safehttp/doc.go:81]: in LIFO order?

File: safehttp/doc.go (hunk 13 above)
Comment (bsiegert) [inline safehttp/doc.go:84]: s/it's/its/

File: safehttp/doc.go (hunk 14 above)
Comment (bsiegert) [inline safehttp/doc.go:87]: The Dispatcher determines

**Review by bsiegert — CHANGES_REQUESTED** [review]

File: safehttp/doc.go (hunk 5 above)
Comment (kele) [reply safehttp/doc.go:46]: It's high level description of the whole package. It also doesn't say "supports exclusively safehtml responses".

File: safehttp/doc.go (hunk 7 above)
Comment (kele) [reply safehttp/doc.go:45]: Added a Warning a few lines below.

File: safehttp/doc.go [hunk 17]
```diff
@@ -12,5 +12,118 @@
...
+//
+// HTTP responses need to be crafted carefully in order to prevent common web
+// vulnerabilities like Cross-site Scripting (XSS). To help with this, we use
+// Safe Responses.
+//
+// Safe Responses are HTTP responses which have been determined to be safe when
+// received by a modern, popular web browser. For example, we consider HTML
+// responses generated using a template system enforcing contextual autoescaping
+// to be safe, e.g. modern Angular or github.com/google/safehtml/template. Read
+// more about contextual autoescaping here
```
Comment (bsiegert) [inline safehttp/doc.go:29]: maybe add a : at end of line

**Review by bsiegert — APPROVED** [review]

**Comment by mtrzos** [comment]
That's really useful. I learned quite a lot about the framework :)

**Pull request merged**